drivers/sensors/paw3222.c

```diff
@@ -0,0 +1,177 @@
+    wait_us(1); // Tncs_lag
+    spi_stop();
+
+    return data;
+}
+
+bool paw3222_init(void) {
+    gpio_set_pin_output(PAW3222_CS_PIN);
+    gpio_set_pin_input(PAW3222_MOTION_PIN);
+
+    // CS must be kept low at power-up stage for at least 1ms
+    gpio_write_pin_low(PAW3222_CS_PIN);
+    wait_ms(10);
+    gpio_write_pin_high(PAW3222_CS_PIN);
+
+    spi_init();
+
+    // reboot
+    paw3222_write(0x06, 0x80);
+    wait_ms(50);
+
+    // initialize
+    paw3222_write(0x09, 0x5A);
+    paw3222_write(0x0D, 0x23);
+    paw3222_write(0x0E, 0x24);
+    paw3222_write(0x19, 0x1C);
+    paw3222_write(0x05, 0xA1);
+    paw3222_write(0x2B, 0x6D);
+    paw3222_write(0x30, 0x2E);
+    paw3222_write(0x5C, 0xDF);
+    paw3222_write(0x7F, 0x01);
+    paw3222_write(0x06, 0x14);
+    paw3222_write(0x31, 0x25);
+    paw3222_write(0x34, 0xC4);
+    paw3222_write(0x36, 0xCC);
+    paw3222_write(0x37, 0x42);
+    paw3222_write(0x38, 0x01);
+    paw3222_write(0x3A, 0x76);
+    paw3222_write(0x3B, 0x34);
+    paw3222_write(0x42, 0x39);
+    paw3222_write(0x43, 0xF2);
+    paw3222_write(0x44, 0x39);
+    paw3222_write(0x45, 0xF0);
+    paw3222_write(0x46, 0x12);
+    paw3222_write(0x47, 0x39);
+    paw3222_write(0x48, 0xE3);
+    paw3222_write(0x49, 0x48);
+    paw3222_write(0x4A, 0xD3);
+    paw3222_write(0x4B, 0x98);
+    paw3222_write(0x64, 0x46);
+    paw3222_write(0x71, 0x28);
+    paw3222_write(0x72, 0x28);
+    paw3222_write(0x7F, 0x00);
+    paw3222_write(0x09, 0x00);
+
+    // read a burst, then discard
+    paw3222_read_burst();
+
+    return paw3222_check_signature();
+}
+
+report_paw3222_t paw3222_read_burst(void) {
+    report_paw3222_t report = {0};
+
+    uint8_t motion = paw3222_read(0x02);
+    if ((motion & MSB1) == MSB1) {
+        // Motion detected
+        uint16_t dx     = (uint16_t)paw3222_read(0x03);
+        uint16_t dy     = (uint16_t)paw3222_read(0x04);
+        uint16_t dxy_hi = (uint16_t)paw3222_read(0x12);
+
+        dx = dx | ((dxy_hi & 0xF0) << 4);
+        dy = dy | ((dxy_hi & 0x0F) << 8);
+
+        report.dx = convert_twoscomp_12(dx);
+        report.dy = convert_twoscomp_12(dy);
+
+        // dprintf("dx is %i, dy is %i\n", report.dx, report.dy);
+    }
+
+    return report;
+}
+
+void paw3222_set_cpi(uint16_t cpi) {
+    uint16_t cpival   = (cpi) < (480) ? (480) : ((cpi) > (4020) ? (4020) : (cpi));
+    uint8_t  cpival_x = (uint8_t)round((float)cpival / 30.0);
+    uint8_t  cpival_y = (uint8_t)round((float)cpival / 29.0);
+
+    paw3222_write(0x09, 0x5A);
+    paw3222_write(0x0D, cpival_x);
+    paw3222_write(0x0E, cpival_y);
+    paw3222_write(0x09, 0x00);
+}
+
+uint16_t paw3222_get_cpi(void) {
+    uint8_t cpival = paw3222_read(0x0D);
+    return (uint16_t)(cpival * 30);
+}
+
+// Convert a 12-bit twos complement binary-represented number into a
+// signed 12-bit integer.
+int16_t convert_twoscomp_12(uint16_t data) {
+    if ((data & 0x800) == 0x800)
+        return -2048 + (data & 0x7FF);
+    else
+        return data;
+}
```

Comment:
```suggestion
```

Reply:
Done in [ff65fd0](https://github.com/qmk/qmk_firmware/pull/25763/commits/ff65fd085f10b82f1a17e161c5da6f06e4f0b39b).